### PR TITLE
Implement node: points from array

### PR DESCRIPTION
**Pull request by msz-rai**

### CODE REVIEW

**Comment by msz-rai** [comment]
Rebased

File: src/graph/Graph.cpp [hunk 1]
```diff
@@ -64,8 +64,11 @@ void Graph::run()
 
 	RGL_DEBUG("Running graph with {} nodes", nodesInExecOrder.size());
 
-	RaytraceNode::Ptr rt = Node::getExactlyOne<RaytraceNode>(nodesInExecOrder);
-	rt->setFields(fieldsToCompute);
+	// If Graph has RaytraceNode set fields to compute
+	if (!Node::filter<RaytraceNode>(nodesInExecOrder).empty()) {
+		RaytraceNode::Ptr rt = Node::getExactlyOne<RaytraceNode>(nodesInExecOrder);
```
Comment (PiotrMrozikRobotec) [inline src/graph/Graph.cpp:69]: To be discussed, what about many raytrace nodes in a single graph.

File: src/graph/FromArrayPointsNode.cpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <graph/NodesCore.hpp>
+#include <gpu/nodeKernels.hpp>
+#include <RGLFields.hpp>
```
Comment (PiotrMrozikRobotec) [inline src/graph/FromArrayPointsNode.cpp:17]: unused includes

File: src/graph/FromArrayPointsNode.cpp [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	fieldData.clear();
+	width = pointCount;
+
+	for (auto&& field : fields) {
+		if (!fieldData.contains(field) && !isDummy(field)) {
+			fieldData.insert({field, VArray::create(field, pointCount)});
+		}
+	}
+
+	VArray::Ptr inputData = VArray::create<char>();
```
Comment (PiotrMrozikRobotec) [inline src/graph/FromArrayPointsNode.cpp:31]: Explicit creation method -> use auto

File: test/src/graphTest.cpp [hunk 4]
```diff
@@ -220,6 +220,64 @@ TEST_F(Graph, FormatNodeResults)
 	}
 }
 
+TEST_F(Graph, UsePointsFromArray)
```
Comment (PiotrMrozikRobotec) [inline test/src/graphTest.cpp:223]: Offtop:
`EXPECT_RGL_SUCCESS(rgl_graph_write_pcd_file(temporalMerge, "two_boxes_temporal_merge.pcd"));` in 164:
shouldnt be in `#ifdef RGL_BUILD_PCL_EXTENSION` ?

File: test/src/graphTest.cpp [hunk 5]
```diff
@@ -220,6 +220,64 @@ TEST_F(Graph, FormatNodeResults)
...
+
+	std::vector<rgl_field_t> pointFields = {
+		XYZ_F32,
+		PADDING_32,
+		IS_HIT_I32,
+		INTENSITY_F32
+	};
+
+	std::vector<PointStruct> inPoints;
+	inPoints.push_back(PointStruct { .xyz = {1, 2, 3}, .isHit = 1, .intensity = 100 } );
```
Comment (PiotrMrozikRobotec) [inline test/src/graphTest.cpp:243]: My suggestion is to change the test to TEST_P and parametrize it with some corner cases.
0 elements, 1 element, 1241251 elems...

Comment (PiotrMrozikRobotec) [reply test/src/graphTest.cpp:243]: But, we can leave it, and I will rewrite it with the next graph_test task.

File: src/gpu/nodeKernels.cu [hunk 6]
```diff
@@ -21,12 +21,21 @@
 #include <thrust/device_ptr.h>
 #include <thrust/scan.h>
 
-__global__ void kFormat(size_t pointCount, size_t pointSize, size_t fieldCount, const GPUFieldDesc* fields, char* out)
+__global__ void kFormatSoaToAos(size_t pointCount, size_t pointSize, size_t fieldCount, const GPUFieldDesc *soaInData, char *aosOutData)
 {
 	LIMIT(pointCount);
 	// Implement padding
 	for (size_t i = 0; i < fieldCount; ++i) {
-		memcpy(out + pointSize * tid + fields[i].dstOffset, fields[i].data + fields[i].size * tid, fields[i].size);
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply src/gpu/nodeKernels.cu:29]: It wouldn't speed up anything. The point count is almost always greater than the number of threads.
@prybicki what do you think?

File: src/graph/Graph.cpp (hunk 1 above)
Comment (msz-rai) [reply src/graph/Graph.cpp:69]: Great catch. I'd apply this change in another PR.

File: test/src/graphTest.cpp (hunk 5 above)
Comment (msz-rai) [reply test/src/graphTest.cpp:243]: Good idea. Let's rewrite it in the new task.

File: src/GPUFieldDescBuilder.hpp [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+					if constexpr (requires { input->fieldData[field]; }) {
+						(*gpuFields)[gpuFieldIdx].writeDataPtr = static_cast<char*>(input->fieldData[field]->getWritePtr(MemLoc::Device));
+					} else {
+						throw InvalidAPIObject("attempted to get writable pointer to field that is not available.");
+					}
+				}
+				++gpuFieldIdx;
+			}
+			offset += getFieldSize(field);
+		}
```
Comment (prybicki) [inline src/GPUFieldDescBuilder.hpp:58]: If possible, try to make it a bit simpler. E.g. use early continue `if (isDummy(field)).` Don't forget to `offset += getFieldSize(field);` in both branches.

File: src/gpu/GPUFieldDesc.hpp [hunk 8]
```diff
@@ -17,7 +17,8 @@
 // Used to send e.g. formatting request to GPU
 struct GPUFieldDesc
 {
-	const char* data;
+	const char* readDataPtr;
+	char* writeDataPtr;
```
Comment (prybicki) [inline src/gpu/GPUFieldDesc.hpp:21]: I'd appreciate an union here with an appropriate comment.

File: src/gpu/nodeKernels.cu [hunk 9]
```diff
@@ -21,12 +21,21 @@
...
+	}
+}
+
+__global__ void kFormatAosToSoa(size_t pointCount, size_t pointSize, size_t fieldCount, const char* aosInData, GPUFieldDesc* soaOutData)
+{
+	LIMIT(pointCount);
+	// Implement padding
+	for (size_t i = 0; i < fieldCount; ++i) {
+		memcpy(soaOutData[i].writeDataPtr + soaOutData[i].size * tid, aosInData + pointSize * tid + soaOutData[i].dstOffset, soaOutData[i].size);
 	}
```
Comment (prybicki) [inline src/gpu/nodeKernels.cu:39]: Hmm, I was boggled by this comment `// Implement padding` here and in SoaToAos. IMHO, it should work because we increase the offset for dummy (padding) fields.

File: src/graph/FromArrayPointsNode.cpp [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <graph/NodesCore.hpp>
+
+void FromArrayPointsNode::setParameters(const void* points, size_t pointCount, const std::vector<rgl_field_t>& fields)
+{
+	// TODO(msz-rai): Optimize memory allocation. Do not clear all fields.
```
Comment (prybicki) [inline src/graph/FromArrayPointsNode.cpp:19]: I don't mind clearing all fields. It's a rare operation, optimizing here is premature IMO. CC @PiotrMrozikRobotec

File: src/graph/Graph.cpp (hunk 1 above)
Comment (prybicki) [reply src/graph/Graph.cpp:69]: IMHO having a single RaytraceNode is a reasonable assumption allowing for some simplifications.

**Review by prybicki — COMMENTED** [review]
Good job!

File: src/graph/Interfaces.hpp [hunk 11]
```diff
@@ -78,6 +78,23 @@ struct IPointsNode
...
+	// Data getters
+	VArray::ConstPtr getFieldData(rgl_field_t field, cudaStream_t stream) const override
+	{ return std::const_pointer_cast<const VArray>(fieldData.at(field)); }
+
+protected:
+	std::unordered_map<rgl_field_t, VArray::Ptr> fieldData;
+
+	friend struct GPUFieldDescBuilder;  // To access fieldData
+};
+
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply src/graph/Interfaces.hpp:97]: The reason for this interface is to add `friend GPUFieldDescBuilder` and pass this interface to that builder (writable case). In the builder, we have two pubic methods (non-templated) where the developer knows what type of nodes he can use within that methods.
I could add `friend` to concrete classes, but the builder method must have had template and no explicit information about what node he can manage.

File: src/graph/NodesCore.hpp [hunk 12]
```diff
@@ -290,6 +288,26 @@ struct TemporalMergePointsNode : Node, IPointsNodeSingleInput
...
+	// Point cloud description
+	bool isDense() const override { return false; }
+	size_t getWidth() const override { return width; }
+	size_t getHeight() const override { return 1; }
+
+	Mat3x4f getLookAtOriginTransform() const override { return Mat3x4f::identity(); }
+
+private:
+	size_t width;
+};
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply src/graph/NodesCore.hpp:309]: This could be implemented in `IPointsSourceNode` as: take any VArray from fieldData and return his elemCount.

File: src/graph/Interfaces.hpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (msz-rai) [reply src/graph/Interfaces.hpp:97]: Also, it opens easy implementation for other IPointsSourceNode. e.g., points from a file, points from rosbag, etc.

File: src/graph/NodesCore.hpp [hunk 13]
```diff
@@ -290,6 +288,26 @@ struct TemporalMergePointsNode : Node, IPointsNodeSingleInput
...
+	// Node
+	void validate() override;
+	void schedule(cudaStream_t stream) override {}
+
+	// Point cloud description
+	bool isDense() const override { return false; }
+	size_t getWidth() const override { return width; }
+	size_t getHeight() const override { return 1; }
+
+	Mat3x4f getLookAtOriginTransform() const override { return Mat3x4f::identity(); }
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply src/graph/NodesCore.hpp:305]: Fixed

File: src/graph/RaytraceNode.cpp [hunk 14]
```diff
@@ -69,10 +71,19 @@ void RaytraceNode::schedule(cudaStream_t stream)
...
 		if (!fieldData.contains(field)) {
 			fieldData.insert({field, VArray::create(field)});
 		}
+		if (oldKeys.contains(field)) {
+			oldKeys.erase(field);
+		}
+	}
+
+	for (auto&& keyToErase : oldKeys) {
+		fieldData.erase(keyToErase);
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply src/graph/RaytraceNode.cpp:87]: 1. OK, c++20 is very helpful.
2. Rare operation, but it could cause annoying lag when changing graphs in runtime. Common case: developing simulation in game engine.

File: src/gpu/GPUFieldDesc.hpp (hunk 8 above)
Comment (msz-rai) [reply src/gpu/GPUFieldDesc.hpp:21]: A union turns out to be unsafe. I will provide an appropriate comment only.

File: src/gpu/nodeKernels.cu (hunk 9 above)
Comment (msz-rai) [reply src/gpu/nodeKernels.cu:39]: This is a comment from the previous implementation. For me either, padding is already implemented. Could some optimizations be applied? If nothing comes to your head, I will remove this comment.

File: src/graph/FromArrayPointsNode.cpp (hunk 10 above)
Comment (msz-rai) [reply src/graph/FromArrayPointsNode.cpp:19]: I could imagine a case where we would like to test the sequence of point clouds. With the increasing number of tests, it would be nice to reduce execution time.

File: src/graph/Interfaces.hpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (msz-rai) [reply src/graph/Interfaces.hpp]: Discussed offline - I have removed the new interface.
